src/navigation/path.cpp

```diff
@@ -0,0 +1,472 @@
+
+    if (distanceToEndNode < distanceToStartNode) {
+        closestPos = endNodePos;
+        closestNode = _end.node();
+    }
+
+    const double distanceToClosestNode = glm::distance(closestPos, prevPose.position);
+    double speed = distanceToClosestNode;
+
+    // Dampen speed in beginning of path
+    double startUpDistance = 2.0 * _start.node()->boundingSphere();
+    if (startUpDistance < Epsilon) { // zero bounding sphere
+        startUpDistance = glm::distance(_start.position(), startNodePos);
+    }
+
+    if (traveledDistance < startUpDistance) {
+        speed *= traveledDistance / startUpDistance + 0.01;
+    }
+
+    // Dampen speed in end of path
+    // Note: this leads to problems when the full length of the path is really big
+    double closeUpDistance = 2.0 * _end.node()->boundingSphere();
+    if (closeUpDistance < Epsilon) { // zero bounding sphere
+        closeUpDistance = glm::distance(_end.position(), endNodePos);
+    }
+
+    if (traveledDistance > (pathLength() - closeUpDistance)) {
+        const double remainingDistance = pathLength() - traveledDistance;
+        speed *= remainingDistance / closeUpDistance + 0.01;
+    }
+
+    // TODO: also dampen speed based on curvature, or make sure the curve has a rounder shape
+
+    // TODO: check for when path is shorter than the starUpDistance or closeUpDistance variables
+
+    return _speedFactorFromDuration * speed;
+}
+
+Waypoint waypointFromCamera() {
+    Camera* camera = global::navigationHandler->camera();
+    const glm::dvec3 pos = camera->positionVec3();
+    const glm::dquat rot = camera->rotationQuaternion();
+    const std::string node = global::navigationHandler->anchorNode()->identifier();
+    return Waypoint{ pos, rot, node };
+}
+
+SceneGraphNode* findNodeNearTarget(const SceneGraphNode* node) {
+    const std::vector<SceneGraphNode*>& relevantNodes =
+        global::navigationHandler->pathNavigator().relevantNodes();
+
+    for (SceneGraphNode* n : relevantNodes) {
+        if (n->identifier() == node->identifier()) {
+            continue;
+        }
+
+        constexpr const float proximityRadiusFactor = 3.f;
+
+        const float bs = static_cast<float>(n->boundingSphere());
+        const float proximityRadius = proximityRadiusFactor * bs;
+        const glm::dvec3 posInModelCoords =
+            glm::inverse(n->modelTransform()) * glm::dvec4(node->worldPosition(), 1.0);
+
+        bool isClose = collision::isPointInsideSphere(
+            posInModelCoords,
+            glm::dvec3(0.0, 0.0, 0.0),
+            proximityRadius
+        );
+
+        if (isClose) {
+            return n;
+        }
+    }
+
+    return nullptr;
+}
+
+// Compute a target position close to the specified target node, using knowledge of
+// the start point and a desired distance from the node's center
+glm::dvec3 computeGoodStepDirection(const SceneGraphNode* targetNode,
+                                     const Waypoint& startPoint)
+{
```

Comment:
`const`

Reply:
This is actually not a member function, but just a helper function to the `createPathFromDictionary` function below. 

Wasn't sure how to structure this part of the code nicely, but many of these functions actually aren't relevant to the `Path` class itself or need to be exposed anywhere else. So, just kept them as independent functions in the cpp file and did not include them in the header file

Is this an ok solution? Open to suggestions on how to make it nicer